saviSimulation/savi/jason_processing/behaviour/UxVBehavior.java

```diff
@@ -0,0 +1,240 @@
+
+                //calculate distance
+                double azimuth = polar.get(Geometry.AZIMUTH);
+                double elevation = polar.get(Geometry.ELEVATION);
+                double dist = polar.get(Geometry.DISTANCE);
+                if ((azimuth < Math.PI / 2. || azimuth > 3 * Math.PI / 2.) && (dist < perceptionDistance)) {
+                    //it's visible
+                    detectedItems.add(new CameraPerception(wo.getType(), this.time, azimuth, elevation, dist, wo.getPixels() / 2));
+                    visibleItems.add(new CameraPerception(wo.getType(), this.time, azimuth, elevation, dist, wo.getPixels() / 2));
+                }
+            }
+        }
+
+        //remove objects covered by others on the visualization
+        for (CameraPerception di : detectedItems) {
+            for (int i = 0; i < visibleItems.size(); i++) {
+                //to calculate visual angle covered by the object
+                distance = di.getParameters().get(2);
+                //angle deviation from centroid = radius
+                oposite = di.getParameters().get(3);
+                //math to calculate angle cover
+                tan = oposite / distance;
+                angle = Math.abs(Math.atan(tan));
+
+                //if object is covered by di remove
+                //if is covered by azimuth angle
+                if (di.getParameters().get(0) + angle > visibleItems.get(i).getParameters().get(0) && di.getParameters().get(0) - angle < visibleItems.get(i).getParameters().get(0)) {
+                    //if it is covered by elevation angle
+                    if (di.getParameters().get(1) + angle > visibleItems.get(i).getParameters().get(1) && di.getParameters().get(1) - angle < visibleItems.get(i).getParameters().get(1)) {
+                        //if it's at a mayor distance
+                        if (di.getParameters().get(2) < visibleItems.get(i).getParameters().get(2)) {
+                            visibleItems.remove(visibleItems.get(i));
+                        }
+                    }
+                }
+            }
+        }
+
+        return visibleItems;
+    }
+
+    /**
+     * Adds to the current compass angle and normalizes the value between 0 and 2 * PI
+     * @param val
+     */
+    protected void addNormalizedCompassAngle(double val) {
+        this.compasAngle += val;
+        double twoPi = 2 * Math.PI;
+
+        if (this.compasAngle < 0)
+            this.compasAngle += twoPi;
+        if (this.compasAngle > twoPi)
+            this.compasAngle -= twoPi;
+    }
+
+    /**
+     * Creates the default actions for a UxV agent
+     */
+    protected void createActionMap() {
+        actionMap.put("turn(left)", (() -> this.addNormalizedCompassAngle(-Math.PI / 16.0)));
+        actionMap.put("turn(right)", (() -> this.addNormalizedCompassAngle(Math.PI / 16.0)));
+        actionMap.put("thrust(on)", (() -> this.speedVector.set(SPEED, SPEED)));
```

Comment:
I don't think this is the semantics of thrust(on)
with this the agents always move in the direction (1,1)

Reply:
there was a speedValue (scalar), and the speed vector was in fact defined as (in polar coordinates) angle = compassAngle and magnitude = speedvalue.

Reply:
have you tested? Because I don't think this would be working properly.

Reply:
I have tested this and it seems to be working just fine!

I replaced speedValue with speedVector so that we could express velocity components in all directions (x, y, z). This made cleaning up the code a lot easier since it allowed us to capture the logic of the UaV and UgV using one vector. When deriving the new position of the agent, it should result in the same position as using the scalar (speedValue). 

If you think the vector is unnecessary, I can revert it to just utilize speedValue! It might be good to add some automated tests (unit tests) so that these kinds of changes won't break the logic. I will see what I can do.

Reply:
see comment at line 102